### PR TITLE
[nn.utils] scale_grad_ with for_each

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150033

Distributed workloads like torch tune often have to do scaling gradients before loss computation.
https://github.com/pytorch/torchtune/blob/main/torchtune/training/_grad_scaler.py#L11

Adding `scale_grad_` that allows to utilize for_each for this, groupping by device, dtype.

Groupping function implemented in python to be dynamo traceable.

Compiled graphs:
```
    def forward(self, arg0_1: "f32[10, 10][10, 1]cpu", arg1_1: "f32[10][1]cpu", arg2_1: "f32[10, 10][10, 1]cuda:0", arg3_1: "f32[10][1]cuda:0"):
         # File: /data/users/ivankobzarev/a/pytorch/test/functorch/test_aotdispatch.py:6627 in fn, code: scale_grad_(test_model.parameters(), torch.tensor(0.5), foreach=True)
        _tensor_constant0 = self._tensor_constant0
        lift_fresh_copy: "f32[][]cpu" = torch.ops.aten.lift_fresh_copy.default(_tensor_constant0);  _tensor_constant0 = None

         # File: /data/users/ivankobzarev/a/pytorch/torch/nn/utils/scale_grad.py:56 in _scale_grad_, code: torch._foreach_mul_(device_grads, scaler.to(device))
        _foreach_mul = torch.ops.aten._foreach_mul.Tensor([arg0_1, arg1_1], lift_fresh_copy)
        getitem: "f32[10, 10][10, 1]cpu" = _foreach_mul[0]
        getitem_1: "f32[10][1]cpu" = _foreach_mul[1];  _foreach_mul = None
        device_put: "f32[][]cuda:0" = torch.ops.prims.device_put.default(lift_fresh_copy, device(type='cuda', index=0));  lift_fresh_copy = None
        convert_element_type: "f32[][]cuda:0" = torch.ops.prims.convert_element_type.default(device_put, torch.float32);  device_put = None
        _foreach_mul_1 = torch.ops.aten._foreach_mul.Tensor([arg2_1, arg3_1], convert_element_type);  convert_element_type = None
        getitem_2: "f32[10, 10][10, 1]cuda:0" = _foreach_mul_1[0]
        getitem_3: "f32[10][1]cuda:0" = _foreach_mul_1[1];  _foreach_mul_1 = None
        copy_: "f32[10, 10][10, 1]cpu" = torch.ops.aten.copy_.default(arg0_1, getitem);  arg0_1 = getitem = copy_ = None
        copy__1: "f32[10][1]cpu" = torch.ops.aten.copy_.default(arg1_1, getitem_1);  arg1_1 = getitem_1 = copy__1 = None
        copy__2: "f32[10, 10][10, 1]cuda:0" = torch.ops.aten.copy_.default(arg2_1, getitem_2);  arg2_1 = getitem_2 = copy__2 = None
        copy__3: "f32[10][1]cuda:0" = torch.ops.aten.copy_.default(arg3_1, getitem_3);  arg3_1 = getitem_3 = copy__3 = None
        return ()
```

### CODE REVIEW

File: torch/nn/utils/scale_grad.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        None
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    else:
+        parameters = list(parameters)
+    _scale_grad_(parameters, scaler, foreach)
+
+
+def _group_tensors_by_device_and_dtype(
```
Comment (janeyx99) [inline torch/nn/utils/scale_grad.py:45]: why redefine this vs use the C++ version?

**Review by janeyx99 — COMMENTED** [review]
I briefly scanned, and I have a few questions about the overall ask:

This use case sounds like the GradScaler we already have in the codebase, but just simpler. https://github.com/pytorch/pytorch/blob/532530be34bf396cc632b5c8d95253953e3f7717/torch/amp/grad_scaler.py Instead of having 2 places in our codebase for one feature, we should reuse/refactor as much as possible. Taking a step back, I'm curious why people want to scale their grads manually--is it not for AMP?

**Comment by IvanKobzarev** [comment]
> I briefly scanned, and I have a few questions about the overall ask:
> 
> This use case sounds like the GradScaler we already have in the codebase, but just simpler. https://github.com/pytorch/pytorch/blob/532530be34bf396cc632b5c8d95253953e3f7717/torch/amp/grad_scaler.py Instead of having 2 places in our codebase for one feature, we should reuse/refactor as much as possible. Taking a step back, I'm curious why people want to scale their grads manually--is it not for AMP?

Thanks for pointing at GradScaler. Do you know if it is compilable and will go through for_each ?

File: torch/nn/utils/scale_grad.py (hunk 1 above)
Comment (janeyx99) [reply torch/nn/utils/scale_grad.py:45]: cc @mlazos who handled this back when working on optimizers--if i remember correctly, it was faster to have the compiler do the grouping?

**Comment by janeyx99** [comment]
@IvanKobzarev I'm not sure what compilable means here. For foreach, the unscale goes through: https://github.com/pytorch/pytorch/blob/main/torch/amp/grad_scaler.py#L283

File: torch/nn/utils/scale_grad.py (hunk 1 above)
Comment (mlazos) [reply torch/nn/utils/scale_grad.py:45]: sorry I missed this, yes I had dynamo do this during tracing, although if you're not compiling, I don't think the eager foreach ops support mixed devices.
